techmatters/hrm
# accountSid part of PK + Review & Enhance HRM DB indexes

## Description
This PR:
- Updates test dump to the state previous to this PR.
- Adds migration to have `accountSid` as part of relevant of PK:
  - Adds `accountSid` column as part of the PK of `Cases`, `Contacts`, `CSAMReports` and `PostSurveys` tables.
  - Adds `accountSid` column to `CaseSections` table, and back-fills it with the value present in the parent case of each record.
  - Adds `accountSid` to be taken into consideration for the relevant queries on `Cases` and `CaseSections` queries. This change is skipped for the rest of the tables because the "controllers" are already taking into consideration and filling the `accountSid` with the corresponding value (or empty string as default. Q: Should we review and change this "empty string default" given that now we have `accountSid` as part of the PK of our tables?).
  - Updates the following FKs: `Contacts` reference to `Cases`, `CaseSections` reference `Cases`, `CSAMReports` reference to `Contacts`.
  - Adjusts CI to the above changes.
- Adds migration that creates new indexes:
  - As [Postgres does not automatically creates indexes on FKs](https://www.postgresql.org/docs/current/ddl-constraints.html#:~:text=A%20foreign%20key%20must,on%20the%20referencing%20columns.), this PR includes indexes matching all of our FK relationships (e.g. index on `caseId`-`accountSid` on `CaseSections` table, to improve queries like "retrieving all the case sections for a given case").
  - Indexes for `Cases` and `Contacts` tables on `accountSid` column.
- Adds two "utility queries" that I found useful while working on this, 
  - `fillDB()` function, used to create records for most of our tables (useful to test cost of queries). Can be created via CLI or PGAdmin and then called like `SELECT fillDB()`.
  -  `selectSingleCaseByIdSql` query, that's just one of the queries we use in our app. Useful when used with `EXPLAIN ANALIZE` to see the difference on the cost of queries before and after the creation of the indexes.

Some results I got before the creation of indexes

![Screenshot from 2022-04-29 17-14-19](https://user-images.githubusercontent.com/15805319/166072183-8dd7d929-d71b-4fd9-ae68-ea13e8f53d92.png)

and after they are created

![Screenshot from 2022-04-29 17-14-48](https://user-images.githubusercontent.com/15805319/166072285-896cf1cc-2db9-44ab-bb85-b1b45fa37a21.png)

For what I could observe, planning and execution costs vary but the ones with indexes are consistently lower than the same queries without the indexes (so they are worth I think :smile:).

### Checklist
- [ ] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1084)

### Verification steps
- [Optional but recommended] Start the test containerized test DB.
- `➜ POSTGRES_PORT=5433 API_KEY=anything USE_OPEN_PERMISSIONS=true npm start` so the migrations are executed against the DB (omit the env vars if not using container).
- Check that everything still works.
- Maybe test the costs of some queries with and without the indexes (you can `➜ POSTGRES_PORT=5433 API_KEY=anything USE_OPEN_PERMISSIONS=true npm run db:undo-last` to remove the indexes, as they are isolated in the latest migration).
